Documentation/General/emergentArchitectureDesign/emergentArchitectureDesign.tex

```diff
@@ -191,3 +191,18 @@ \section{Concurrency}
 These seperate threads will get a specific task assigned, which has his local data to compute on.
 Deadlocks will be prevented because the spawned threads will be joined before acquiring locks in synchronized statements.
 This will happen when we persist the algorithm output in the database.
+
+\section{Scalability}
+Concurrent clients are one thing, but we also have to scale the algorithms for growth.
+Initially, we planned on a nearest neighbours algorithm to match songs to rooms.
+Later on, we found out that while this works well for small datasets, finding the nearest neighbours in a system of a few million songs can prove to be a lot harder.
+The mission is clear: Find a way for the system to scale further.
+We went over several iterations of the system.
+At first, we had the nearest neighbours algorithm.
+Once we realised the issue here, we went for an ANN\footnote{Approximate nearest neighbour} algorithm to be able to use the nearest neighbour algorithm.
+This initially was planned to be a \gls{KD-Tree}.
+This still had issues with memory usage (O(N) complexity).
```

Comment:
However, this would cause memory issues due to a O(N) complexity of the algorithm.